cli: add design guidelines

daniel-weisse opened the pull request:
<!--
Thank you for your contribution!

For more information check our contributors guide CONTRIBUTING.md (link below text box).

NOTE: This template is a guideline to help you to provide meaningful information for reviewers.
Feel free to edit, complete or extend this list while the PR is open.
-->
### Context
In preparation of refactoring I went over the CLI packages and wrote out some ideas and guidelines for the CLI.
The document aims to explain the design goals of packages in the CLI, not necessarily their function.
Any suggestions are welcome.

Especially from the codeowners of the packages I would like to hear:
* Does the suggested spec align with your view of the package
* Would you like to add any information on the package or change it in some way?

### Proposed change(s)
- Add CLI design guide
- Remove iamid package in favor of declaring the struct in the `cloudcmd` package.

<!-- (uncomment if applicable)
### Related issue
- link to the issue
-->

<!-- (uncomment if applicable)
### Additional info
- Any additional information or context
-->

### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x], or check after submitting. -->
<!-- more information in dev-docs/workflows/pull-request.md -->
- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [ ] Add labels (e.g., for changelog category)
- [ ] Is PR title adequate for changelog?
- [ ] Link to Milestone

derpsteb commented on cli/README.md:
```diff
@@ -0,0 +1,154 @@
+
+It is used by the [`cmd` package](#cmd) to handle creation of cloud resources and other CSP specific interactions.
+
+### [`clusterid`](./internal/)
```
```suggestion
### [`clusterid`](./internal/clusterid)
```

derpsteb commented on cli/README.md:
```diff
@@ -0,0 +1,154 @@
+### [`helm`](./internal/helm)
+
+Helm is used to manage Kubernetes deployments of a Constellation cluster.
+The package implements functionality to install, update, and manage Helm charts.
```
```suggestion
The package implements functionality to install, update, and manage Helm charts.
The package includes cloud-provider specific implementations, as charts need to be rendered differently for each CSP.
Exported functions should be CSP agnostic and take a cloudprovider argument, like described for pkg `cloudcmd`.
```

elchead replied:
I think `helm` can be CSP specific since otherwise the `cloudcmd` pkg would have to know about helm charts which is worse.

derpsteb commented on cli/README.md:
```diff
@@ -0,0 +1,154 @@
+The package implements functionality to install, update, and manage Helm charts.
+
+The charts themselves are embedded in a built CLI binary, and values are dynamically updated depending on configuration.
+The charts can be found in [`cli/internal/helm/charts/`](./internal/helm/charts/).
```
```suggestion
The charts can be found in [`cli/internal/helm/charts/`](./internal/helm/charts/).
Values should be added in the chart's `values.yaml` files, if they are static i.e. don't depend on user input.
```

derpsteb commented on cli/README.md:
```diff
@@ -0,0 +1,154 @@
+Terraform is used to create cloud resources and IAM resources required for Constellation clusters.
+The package implements functionality to create, update, delete, and view said resources.
+
+The Terraform files are embedded in a built CLI binary, and variables are dynamically update depending on configuration.
```
```suggestion
The Terraform files are embedded in the CLI binary, and variables are dynamically updated depending on configuration.
```

derpsteb commented on cli/README.md:
```diff
@@ -0,0 +1,154 @@
+The Terraform files are embedded in a built CLI binary, and variables are dynamically update depending on configuration.
+The files can be found in [`cli/internal/terraform/terraform/`](./internal/terraform/terraform/).
+
+Functions of this package should be kept mostly cloud agnostic (There should be no "CreateAzureCluster" function),
```
```suggestion
Functions of this package should be kept CSP agnostic (there should be no "CreateAzureCluster" function),
```

If you want to keep the "mostly", could you specify what that means?

derpsteb commented:
Maybe @katexochen should have a look as well if you didn't discuss with him already. Maybe he has some resources for pkg design/when (not) to add one.

katexochen commented:
Thanks for writing this down. lgtm in principal, but

- IMO the package specific descriptions belong in the pkg comment, not into a readme. Duplicating this information here in addition will likely only add to the problem, as we already fail to keep pkg comments up to date.
- The remaining parts aren't specific to the cli/cmd package, but general information on code design. IMO these belong into the dev-docs.

And some random thoughts while reading: Renaming helm->helmcmd and kubernetes->kubecmd might help to clarify the intent of these pkgs.

katexochen commented:
> Maybe he has some resources for pkg design/when (not) to add one.

No, sorry, none come to my mind right now.

elchead commented on cli/README.md:
```diff
@@ -0,0 +1,154 @@
+
+[`cli/cmd/`](./cmd/) defines the root command of the CLI. Any changes concerning the root command, or additions like persistent flags (flags that can be accessed by all subcommands) should be added here.
+
+TODO: Discuss: Root command could be added to `cli/internal/cmd/`
```
since root is not intended to be reused and is application code, it seems fine in its current place

elchead commented on cli/README.md:
```diff
@@ -0,0 +1,154 @@
+
+All filepaths handled by the CLI code should originate from here.
+Common filepaths are defined as constants in [`internal/constants`](../internal/constants/).
+To generate workspace correct filepaths for printing, use the functions from [`workspace.go`](./internal/cmd/workspace.go) // TODO: Discuss if we should move this to its own package
```
own pkg makes sense to me

daniel-weisse commented:
Cleaned up and implemented suggestions:

* Moved the "Adding to a package/Adding new packages" to `dev-docs/conventions.md`
* Moved the package design goals from a separate README directly to the package documentation
* Moved the workspace path functions to their own (sub) package of `cli/internal/cmd/workspace`

elchead commented on cli/internal/cloudcmd/iam.go:
```diff
@@ -201,23 +200,51 @@ func (c *IAMCreator) createAWS(ctx context.Context, cl tfIAMClient, opts *IAMCon
+	// CloudProvider is the cloud provider of the cluster.
+	CloudProvider cloudprovider.Provider `json:"cloudprovider,omitempty"`
+
+	GCPOutput   GCPIAMOutput   `json:"gcpOutput,omitempty"`
```
should we make the fields optional through a ptr?

daniel-weisse replied:
They weren't before. This struct also is never written to disk, so making fields optional doesn't change much, but requires us to make additional safety checks (the json tags also don't really matter here).

elchead commented on cli/internal/terraform/terraform.go:
```diff
@@ -5,12 +5,15 @@ SPDX-License-Identifier: AGPL-3.0-only
 
 /*
-Package terraform handles creation/destruction of a Constellation cluster using Terraform.
+Package terraform handles creation/destruction of a cloud and IAM resources required by Constellation using Terraform.
```
```suggestion
Package terraform handles creation/destruction of cloud and IAM resources required by Constellation using Terraform.
```

elchead reviewed (approved):
thanks, these are good improvements and will be especially helpul for new onboardings :)

katexochen commented on cli/internal/cmd/workspace/workspace.go:
```diff
@@ -0,0 +1,65 @@
+*/
+
+/*
+Package workspace provides functions to generate paths relative to the workspace.
```
I don't think I get the reason for this package yet. Can't we integrate this in a file handler or struct? Looks like something that should happen in a more automatic way. `PrefixFileHandler` or so...

daniel-weisse replied:
Do I understand you correctly that you would prefer some interface to handle the prefixing?
E.g. `PrefixHandler` would be set up at the start of a command using the workspace flag.
Then every time we want to print a filepath, instead of calling a distinct function, we instead call, e.g. `PrefixHandler.Path(constants.<SomeFilename>)`, which would return the correct path for the user?

daniel-weisse replied:
Implemented a struct to handle filepath prefixing, can you please review?

katexochen reviewed (approved):
lgtm, thanks. :)

3u13r commented on cli/internal/cmd/workspace/workspace.go:
```diff
@@ -0,0 +1,65 @@
+Users may override the default workspace using the --workspace flag.
+The default workspace is the current working directory.
+
+The functions defined here return paths relative to the set workspace,
+and should be used when printing the path to the user.
+
+The MUST not be used when accessing files, as the workspace is changed
+using os.Chdir() before the command is executed.
```
`return paths relative to the set workspace` this is not true, as far I understand it. They return the path to the actual pwd (before the -C execution). 

To avoid misuse, should we omit this functionality entirely and print the paths in relation to the set workdir? How do other tools behave here? I always understood `-C` as literally `cd` into the folder but if the program returns, I want to be back in the original path. I'd assume that e.g. `make` does not "normalize" the paths as we do here.

daniel-weisse replied:
>I always understood -C as literally cd into the folder

Correct. One exit, you will be back where you started.
However, we have multiple cases where we print filepaths to the user for 
For example, after a successful init we print the path to the kubeconfig file:

```
Constellation ID file written to <workspace>/constellation-admin.conf
You can now connect to your cluster by executing:
    export KUBECONFIG="$PWD/<workspace>/constellation-admin.conf
```
If we omit this package, those paths will not work correctly.

(Looking at this again, this path may also not be printed correctly when one provides an absolute path. Should be easy to fix though.)

3u13r replied:
> However, we have multiple cases where we print filepaths to the user for

What are the other cases? Can we simply always print the absolute path?

daniel-weisse replied:
> What are the other cases?

Usually any time we create or update files, e.g. during config generate, and upgrade apply/check. It is also required for printing error messages to the user.
We also print a lot of filenames for debug statements.

> Can we simply always print the absolute path?

In that case we would always have to call `filepath.Abs`, so there is no real advantage gained, other than not parsing the workspace flag, with the downside of paths looking less pretty.

I have also implemented the suggestion from https://github.com/edgelesssys/constellation/pull/2181#discussion_r1287386166, which should make usage of this package a lot easier.

derpsteb reviewed (approved):
:muscle:

katexochen commented on cli/internal/cmd/pathprefix/pathprefix.go:
```diff
@@ -0,0 +1,38 @@
+The functions defined here should be used when printing any filepath to the user,
+as they might otherwise be incorrect if the user has changed the workspace.
+
+The MUST not be used when accessing files, as the workspace is changed
```
```suggestion
The prefixer MUST not be used when accessing files, as the workspace is changed
```
